CMakeLists.txt

```diff
@@ -263,145 +264,159 @@ endif ()
 
 set (FFTW_VERSION 3.3.7)
 
-set (PREC_SUFFIX)
-if (ENABLE_FLOAT)
-  set (FFTW_SINGLE TRUE)
-  set (BENCHFFT_SINGLE TRUE)
-  set (PREC_SUFFIX f)
+if (BUILD_SHARED_LIBS)
+  add_definitions (-DFFTW_DLL)
```

Comment:
`GenerateExportHeader` should be used instead:
* https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html

Reply:
Example:
* https://github.com/forexample/package-example/blob/a5ecaa520466ca286fb31e3fc0753901ae5ff0b4/Foo/CMakeLists.txt#L41-L46
* https://github.com/forexample/package-example/blob/a5ecaa520466ca286fb31e3fc0753901ae5ff0b4/Foo/Source/foo/Bar.hpp#L6-L10

Reply:
This is code I kept from the original `CMakeLists.txt`, so I'm hesitant to change it as it's not really related to the issue this PR is addressing. Or would you argue the change is necessary for clean use within Hunter?

Reply:
As far as I understand this code will be used in installed headers too (can't check because I can't install). Meaning the burden of setting `FFTW_DLL` moved to user. `GenerateExportHeader` designed to resolve all problems.

> Or would you argue the change is necessary for clean use within Hunter?

I'm pretty sure you we will hit this problem while testing.

Reply:
I see, thanks for explaining. I tested this PR on Windows and didn't have any problems regarding export symbols, so I think it's fine this way. [`ffftw.h` lines 78 to 87](https://github.com/FFTW/fftw3/blob/708b202fd593cf1002cf97dce0863e2a438e3720/api/fftw3.h#L78-87) handles this fine as far as I can see:

```cpp
#if defined(FFTW_DLL) && (defined(_WIN32) || defined(__WIN32__))
   /* annoying Windows syntax for shared-library declarations */
#  if defined(COMPILING_FFTW) /* defined in api.h when compiling FFTW */
#    define FFTW_EXTERN extern __declspec(dllexport)
#  else /* user is calling FFTW; import symbol */
#    define FFTW_EXTERN extern __declspec(dllimport)
#  endif
#else
#  define FFTW_EXTERN extern
#endif
```

I can do some further testing. If it turns out not to be sufficient, I'll open a separate PR upstream.

Reply:
> I tested this PR on Windows and didn't have any problems regarding export symbols, so I think it's fine this way

You tested what exactly? You just run build of the project? Have you tried to install it and add it to user's project with `find_package(fftw3 CONFIG REQUIRED)`?

Reply:
> Have you tried to install it and add it to user's project with find_package(fftw3 CONFIG REQUIRED)?

Yes, I successfully built and ran a project that depends on FFTW3 using Hunter on MSVC, so I did 

```cmake
hunter_add_package(fftw3)
find_package(fftw3f REQUIRED)
```

I have the Hunter patches on my local machine and will submit them once this PR is settled.

FYI, I'm a developer on the [LMMS](https://lmms.io/) project (https://github.com/LMMS/lmms). Adding FFTW3 (and some other packages) to Hunter is part of efforts to make LMMS compatible with MSVC. We plan on using Hunter to fetch dependencies on Windows.

Reply:
Also, upstream still supports the autotools build system, so I don't think a  `GenerateExportHeader` patch would be accepted because it could render the two systems incompatible.

Reply:
> hunter_add_package(fftw3)
> find_package(fftw3f REQUIRED)

It should be with `CONFIG`:

```cmake
hunter_add_package(fftw3)
find_package(fftw3f CONFIG REQUIRED)
```

Note that my original question was about shared library (DLL on Windows). Hunter by default build static library, you have to add `CMAKE_ARGS BUILD_SHARED_LIBS=ON`:
* https://github.com/ruslo/hunter/wiki/example.custom.config.id#custom-cmake-options

As a summary it's up to you. I'm just pretty sure this will lead to opaque linker error on user's side.

Reply:
> It should be with CONFIG:

Using `CONFIG` will break compatibility with non-Hunter builds.

> Hunter by default build static library

Ah, I wasn't aware of that. I tried using `BUILD_SHARED_LIBS=ON` but didn't get any linking errors. Anyway, I'll test this on more environments in the context of LMMS and will submit another patch in case I run into problems.

Reply:
> Using CONFIG will break compatibility with non-Hunter builds

I think you're already using `CONFIG` implicilty.